src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,443 @@
+/-
+Copyright (c) 2022 Jon Eugster. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jon Eugster
+-/
+import algebra.char_p.algebra
+import algebra.char_p.local_ring
+import data.pnat.basic
+import ring_theory.ideal.quotient
+import tactic.field_simp
+
+/-!
+# Equal and mixed characteristic
+
+In commutative algebra, some statments are simpler when working over a `ℚ`-algebra `R`, in which
+case one also says that the ring has "equal characteristic zero". A ring that is not a
+`ℚ`-algebra has either positive characteristic or there exists a prime ideal `I ⊂ R` such that
+the quotient `R ⧸ I` has positive characteristic `p > 0`. In this case one speaks of
+"mixed characteristic `(0, p)`", where `p` is only unique if `R` is local.
+
+Examples of mixed characteristic rings are `ℤ` or the `p`-adic integers/numbers.
+
+This file provides the main theorem `split_by_characteristic` that splits any proposition `P` into
+the following three cases:
+
+1) Positive characteristic: `char_p R p` (where `p ≠ 0`)
+2) Equal characteristic zero: `algebra ℚ R`
+3) Mixed characteristic: `mixed_char_zero R p` (where `p` is prime)
+
+## Main definitions
+
+- `mixed_char_zero` : A ring has mixed characteristic `(0, p)` if it has characteristic zero
+  and there exists an ideal such that the quotient `R ⧸ I` has characteristic `p`.
+
+## Main results
+
+- `split_equal_mixed_char` : Split a statement into equal/mixed characteristic zero.
+
+This main theorem has the following three corollaries which include the positive
+characteristic case for convenience:
+
+- `split_by_characteristic` : Generally consider positive char `p ≠ 0`.
+- `split_by_characteristic_domain` : In a domain we can assume that `p` is prime.
+- `split_by_characteristic_local_ring` : In a local ring we can assume that `p` is a prime power.
+
+## TODO
+
+- Relate mixed characteristic in a local ring to p-adic numbers [number_theory.padics].
+
+-/
+
+variables (R : Type*) [comm_ring R]
+
+/-!
+### Mixed characteristic
+-/
+
+/--
+A ring of characteristic zero is of "mixed characteristic `(0, p)`" if there exists an ideal
+such that the quotient `R ⧸ I` has caracteristic `p > 0`.
+-/
+class mixed_char_zero {p : ℕ} : Prop :=
+[to_char_zero : char_zero R]
+(p_pos : p ≠ 0)
+(char_p_quotient : ∃ (I : ideal R), (I ≠ ⊤) ∧ char_p (R ⧸ I) p)
```

Comment:
Linter tells me to make `p` a metavariable, but I'm not fully sure if that's a good idea, since it seems to me that `mixed_char R p` and `mixed_char R q` might both be properties a ring has simultaneously...

Reply:
Now changed in to a `pnat`.